File: aznavrail-react/src/AzNavRailDefaults.ts
/** Compile-time constants that mirror the default dimension and layout values from the Kotlin library. */
export const AzNavRailDefaults = {
  /** Width and height of the app-icon in the rail header, in dp. */
  HeaderIconSize: 48,
  /** Padding around the header icon container, in dp. */
  HeaderPadding: 16,
  /** Total height of the rail header row, in dp. */
  HeaderHeightDp: 56,
  /** Horizontal padding inside the rail's icon column, in dp. */
  RailContentHorizontalPadding: 12,
  /** Vertical gap between consecutive rail buttons, in dp. */
  RailContentVerticalArrangement: 8,
  /** Minimum swipe distance (px) required to trigger undocking the rail into floating mode. */
  SWIPE_THRESHOLD_PX: 50,
  /** Distance (px) from the docked position within which a released floating rail snaps back to the edge. */
  SNAP_BACK_RADIUS_PX: 100,
  /** Default rail width when collapsed to icon-only mode, in dp. */
  CollapsedRailWidth: 100,
  /** Default rail width when expanded to show the full menu, in dp. */
  ExpandedRailWidth: 160,
  /** Standard button size used in the full-size rail, in dp. */
  ButtonWidth: 72,
  /** Smaller button size used when the rail is in compact/shrunk mode, in dp. */
  ShrunkButtonWidth: 56,
  /** Sentinel value for `screenTitle` meaning no title should be shown in the host header. */
  NO_TITLE: 'NO_TITLE',
  /** The colour a rail item falls back to when neither the item nor the host names one. */
  AccentFallback: '#6200ee',
  /**
   * Auto-generated ID for the implicit About (`?`) button appended to the end of the rail when
   * the developer declared none of their own.
   */
  AutoAboutId: 'AZ_AUTO_ABOUT_ID_INTERNAL',
};

/**
 * The library's one motion scale — the TypeScript twin of Kotlin's `AzMotion`.
 *
 * Every transition in the port reads its timing from here. Before this existed the same two
 * numbers — a 60 ms stagger and a 720 ms item duration — were written out as literal defaults in
 * seven separate places, which is how they drifted past the point of being useful: an eight-item
 * drawer took **1.2 seconds** to finish arriving, and because the turnstile entrance starts each
 * item edge-on (and therefore invisible), the panel sat there empty for the first stretch of it.
 *
 * Motion here is a guide, not a toll. It tells the user where a thing came from and then gets out
 * of the way. The durations are drawn from Material 3 Expressive's scale and deliberately sit at
 * its quick end: a cascade the user has to wait out has stopped being a cue and become a delay.
 */
export const AzMotion = {
  /** One item's own entrance or exit. M3 Expressive `medium` — spatial, but on the fast side. */
  ItemDurationMs: 280,
  /**
   * The gap between one item starting and the next. Small enough that the cascade reads as a
   * single gesture rather than a queue.
   */
  ItemStaggerMs: 22,
  /** A container arriving or leaving — a panel, a scrim, a popup. M3 Expressive `effects`. */
  PanelDurationMs: 200,
  /** A layout settling into a new size, e.g. the rail's collapsed/expanded width. */
  SettleDurationMs: 240,
  /** One step of a continuous, non-blocking indicator. */
  IndicatorStepMs: 420,
} as const;

File: aznavrail-cmp/src/commonMain/kotlin/com/hereliesaz/aznavrail/internal/AzNavRailDefaults.kt
package com.hereliesaz.aznavrail.internal

import androidx.compose.ui.unit.dp

/**
 * Default dimension and sentinel-value constants used throughout the AzNavRail rendering system.
 *
 * Ported verbatim from the Android module's `internal/RailState.kt`. Kept in its own file here so it
 * can live in `commonMain` — the Android original shares a file with `AzNavRailLogger` (which uses
 * `android.util.Log`), which is a Tier-2 file expect/actual will unblock later. Because the values
 * are compile-time constants that never diverge across platforms, the small duplication is safe.
 */
internal object AzNavRailDefaults {
    /** Minimum horizontal drag distance (px) before a swipe-to-open gesture is recognised. */
    const val SWIPE_THRESHOLD_PX = 20f
    /** FAB-mode: distance from origin (px) within which the rail snaps back to its docked position. */
    const val SNAP_BACK_RADIUS_PX = 50f
    /** Padding applied to the header row on all sides. */
    val HeaderPadding = 8.dp

    /** Strict unified button width shared by all rail items, app icons, and nested-rail items. */
    val ButtonWidth = 72.dp
    /** Reduced button width used when a vertical nested rail popup is open. */
    val ShrunkButtonWidth = 56.dp

    val HeaderTextSpacer = 8.dp
    val RailContentHorizontalPadding = 4.dp
    /** Vertical gap between rail buttons when not in packed mode. */
    val RailContentVerticalArrangement = 8.dp
    /** Height of the spacer rendered in place of a divider item in the collapsed rail. */
    val RailContentSpacerHeight = 72.dp
    val MenuItemHorizontalPadding = 24.dp
    val MenuItemVerticalPadding = 12.dp
    val FooterDividerHorizontalPadding = 16.dp
    val FooterDividerVerticalPadding = 8.dp
    val FooterSpacerHeight = 12.dp
    /** Fixed height of the header row (matching the standard button width for alignment). */
    val HeaderHeightDp = 72.dp

    /** Sentinel value for `AzNavItem.screenTitle` meaning "do not show a title". */
    const val NO_TITLE = "NO_TITLE_AZ_NAV_RAIL"
    /** Auto-generated ID used for the implicit help button injected when no explicit help item is present. */
    const val AUTO_HELP_ID = "AZ_AUTO_HELP_ID_INTERNAL"
}
